x/rewards/keeper/min_cons_fee.go

```diff
@@ -0,0 +1,65 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/archway-network/archway/pkg"
+	"github.com/archway-network/archway/x/rewards/types"
+)
+
+// UpdateMinConsensusFee calculates and updates the minimum consensus fee if eligible emitting an event.
+func (k Keeper) UpdateMinConsensusFee(ctx sdk.Context, inflationRewards sdk.Coin) {
+	// Prepare and verify inputs
+	if inflationRewards.IsZero() {
+		k.Logger(ctx).Info("Minimum consensus fee update skipped: inflation rewards are zero")
+		return
+	}
+	inflationRewardsAmt := inflationRewards.Amount.ToDec()
+
+	blockGasLimit := pkg.NewDecFromUint64(ctx.BlockGasMeter().Limit())
+	if blockGasLimit.IsZero() {
+		k.Logger(ctx).Info("Minimum consensus fee update skipped: block gas limit is not set")
+		return
+	}
+
+	txFeeRebateRatio := k.TxFeeRebateRatio(ctx)
+
+	// Calculate
+	feeAmt := calculateMinConsensusFeeAmt(inflationRewardsAmt, blockGasLimit, txFeeRebateRatio)
+	if feeAmt.IsZero() {
+		k.Logger(ctx).Info("Minimum consensus fee update skipped: calculated amount is zero")
+		return
+	}
+	feeCoin := sdk.DecCoin{
+		Denom:  inflationRewards.Denom,
+		Amount: feeAmt,
+	}
+
+	// Set and emit event
+	k.state.MinConsensusFee(ctx).SetFee(feeCoin)
+	k.Logger(ctx).Info("Minimum consensus fee update", "fee", feeCoin)
+
+	types.EmitMinConsensusFeeSetEvent(ctx, feeCoin)
+}
+
+// GetMinConsensusFee returns the minimum consensus fee.
+// Fee defines the minimum gas unit price for a transaction to be included in a block.
+func (k Keeper) GetMinConsensusFee(ctx sdk.Context) *sdk.DecCoin {
```

Comment:
It's not common to return sdk.DecCoin pointers. Internal fields are pointers anyways so no memory copy overhead. Then in case it does not exists we can return `sdk.DecCoin, bool`